relations/src/shielder/tangle.rs

```diff
@@ -1,67 +1,81 @@
-//!  2.1 Given a new mangled sequence of `n` elements we squash it `SQUASH_FACTOR` times, i.e. we
-//!      take chunks of length `SQUASH_FACTOR` and reduce them to a single byte by xoring.
+//!  2.1 A new mangled sequence of `n` elements is reduced by summing.
+//!
+//! In some places, where an element turns out to be zero, we replace it by an index-dependent
+//! constant.
 //!
 //! Note, it is **not** hiding like any hashing function.
 //!
-//! This module exposes two implementations of tangling: `tangle` and `tangle_in_field`. They are
+//! This module exposes two implementations of tangling: [tangle] and [tangle_in_circuit]. They are
 //! semantically equivalent, but they just operate on different element types.
 //!
 //! All the index intervals used here are closed-open, i.e. they are in form `[a, b)`, which means
 //! that we consider indices `a`, `a+1`, ..., `b-1`. We also use 0-based indexing.
 
-use ark_r1cs_std::R1CSVar;
+use std::ops::Add;
+
+use ark_ff::{Field, Zero};
+use ark_r1cs_std::fields::FieldVar;
 use ark_relations::r1cs::SynthesisError;
-use ark_std::vec::Vec;
 
-use super::types::ByteVar;
+use crate::{environment::FpVar, CircuitField};
 
 /// Bottom-level chunk length.
 const BASE_LENGTH: usize = 4;
+/// Tangling operates on fixed-sized arrays with 128 elements.
+const EXPAND_TO: usize = 128;
 
-/// Tangle elements of `bytes`.
+/// Entangle `input` into single `FpVar`.
 ///
 /// For circuit use only.
-pub(super) fn tangle_in_field<const SQUASH_FACTOR: usize>(
-    mut bytes: Vec<ByteVar>,
-) -> Result<Vec<ByteVar>, SynthesisError> {
-    let number_of_bytes = bytes.len();
-    _tangle_in_field(&mut bytes, 0, number_of_bytes)?;
-    Ok(bytes
-        .chunks(SQUASH_FACTOR)
-        .map(|chunk| {
-            chunk
-                .iter()
-                .cloned()
-                .reduce(|x, y| x.xor(&y).unwrap())
-                .unwrap()
-        })
-        .collect())
+pub(super) fn tangle_in_circuit(input: &[FpVar]) -> Result<FpVar, SynthesisError> {
+    let mut input_expanded = input
+        .iter()
+        .cycle()
+        .take(EXPAND_TO)
+        .cloned()
+        .collect::<Vec<_>>();
+
+    _tangle_in_circuit(&mut input_expanded, 0, EXPAND_TO)?;
+
+    Ok(input_expanded.into_iter().reduce(|a, b| a.add(b)).unwrap())
+}
+
+fn dezeroize_in_circuit(
+    fp: &FpVar,
+    fallback: impl Into<CircuitField>,
+) -> Result<FpVar, SynthesisError> {
+    let fallback = FpVar::constant(fallback.into());
+    fp.is_zero()?.select(&fallback, fp)
 }
 
 /// Recursive and index-bounded implementation of the first step of the `tangle` procedure.
-fn _tangle_in_field(bytes: &mut [ByteVar], low: usize, high: usize) -> Result<(), SynthesisError> {
-    // Bottom level case: computing suffix sums. We have to do some loop-index boilerplate, because
-    // Rust doesn't support decreasing range iteration.
+fn _tangle_in_circuit(elems: &mut [FpVar], low: usize, high: usize) -> Result<(), SynthesisError> {
```

Comment:
I assume that the underscore at the beginning comes from python convention but in rust this naming is reserved for sth else.  Maybe sth like `tangle_in_circuit_step` or  `tangle_in_circuit_rec`?

Reply:
changed to `do_...`